Manual on the standard TLA+ operators (#339)
informalsystems/apalache: docs/lang/logic.md

non-determinism in TLA+. See [Control Flow and Non-determinism]. In this
section, we only consider the deterministic version.

----------------------------------------------------------------------------

<a name="forallBounded"></a>
### Bounded universal quantifier

**Notation:** `\A x \in S: P`

**LaTeX notation:** ![forall-bounded](./img/forall-bounded.png)

**Arguments:** At least three arguments: a variable name, a set, and an
expression. As usual in TLA+, if the second argument is not a set, the result is
undefined. You can also use multiple variables and tuples, see **Advanced
syntax**. 

**Effect:** This operator evaluates to a Boolean value. We explain
semantics only for a single variable:

 - `\A x \in S: P` evaluates to `TRUE`, if for every element `e` of `S`, the
   expression `P` evaluates to `TRUE` against the binding `[x |-> e]`.
 - Conversely, `\A x \in S: P` evaluates to `FALSE`, if there exists an element
   `e` of `S` that makes the expression `P` evaluate to `FALSE` against the
   binding `[x |-> e]`.

 _Importantly, when `S = {}`, the expression `\A x \ in S: P` evaluates to
 `TRUE`, independently of what is written in `P`. Likewise, when `{x \in S: P}

>  = {}`, the expression `\A x \ in S: P` evaluates to `TRUE`._

Reviewer: ```
Likewise, when `{x \in S: P} = {}`, the expression `\A x \ in S: P` evaluates to `TRUE`.
```
This is not true, take `S == {1}`, `P == x = 2`. While it is true that `{x \in {1}: x = 2}` equals `{}`, the formula `\A x \in {1}: x = 2` is `FALSE`. 
If you want to state quantification as filtering, you have the equivalence between `{x \in S: P} = S` and ``\A x \ in S: P`.

   
**Determinism:** Deterministic.

**Errors:** Pure TLA+ does not restrict the operator arguments.  TLC flags a
model checking error, if `S` is infinite.  Apalache produces a static type
error, if the type of elements of `S` is not compatible in the context of `P`

> when an element of `S` is bound to `x`.

Reviewer: This sentence is too confusing (imprecise).

Author: Fixed

**Advanced syntax:** Instead of a single variable `x`, you can use the tuple
syntax to unpack variables from a Cartesian product, see [Tuples](./tuples.md).
For instance, one can write `\A <<x, y>> \in S: P`. In this case, for every
element `e` of `S`, the variable `x` is bound to `e[1]` and `y` is bound to
`e[2]`. The predicate `P` is evaluated against this binding.

Moreover, instead of introducing one variable, one can quantify over several
sets. For instance, you can write: `\A x \in S, y \in T: P`. This form is
simply syntax sugar for the form with nested quantifiers: `\A x \in S: \A y
\in T: P`.

> 

Reviewer: \+ Similarly, `\A x,y \in S: P` is syntax sugar for `\A x \in S : \A y \in S: P`.

  T2 = frozenset((x, y) for x in [1, 2] for y in [3, 4])
  all(x < y for (x, y) in T2)   # True
```

----------------------------------------------------------------------------

<a name="existsBounded"></a>
### Bounded existential quantifier

**Notation:** `\E x \in S: P`

**LaTeX notation:** ![exists-bounded](./img/exists-bounded.png)

**Arguments:** At least three arguments: a variable name, a set, and an
expression. As usual in TLA+, if the second argument is not a set, the result is
undefined.You can also use multiple variables and tuples, see **Advanced
syntax**. 

**Effect:** This operator evaluates to a Boolean value. We explain
semantics only for a single variable:

 - `\E x \in S: P` evaluates to `TRUE`, if there is an element `e` of `S`
    that makes the expression `P` evaluate to `TRUE` against the binding
    `[x |-> e]`.
 - Conversely, `\E x \in S: P` evaluates to `FALSE`, if for all elements
   `e` of `S`, the expression `P` evaluate to `FALSE` against the
   binding `[x |-> e]`.

 _Importantly, when `S = {}`, the expression `\E x \ in S: P` evaluates to
 `FALSE`, independently of what is written in `P`. Likewise, when `{x \in S: P}

>  = {}`, the expression `\A x \ in S: P` evaluates to `FALSE`._

Reviewer: This is true, but I don't know why you're giving an example for `\A` in the section on `\E`.

Author: It was also wrong. Fixed!

 As you probably have noticed, `\E x \in S: P` is equivalent to `~(\A x \in S:
 P)`, and `\A x \in S: P` is equivalent to `~(\E x \in S: P)`. This is called

>  _duality_ in logic. But take care!  If `P` contains primes inside, `\E x \in

Reviewer: Again, the categorization with "contains primes" is incomplete.

 S: P` may work as a non-deterministic assignment. In this case, duality does
 not work anymore!  See [Control Flow and Non-determinism].
   

> **Determinism:** Deterministic, unless `P` contains primes.

Reviewer: See above.

For the non-deterministic case, see [Control Flow and Non-determinism].

**Errors:** Pure TLA+ does not restrict the operator arguments.  TLC flags a
model checking error, if `S` is infinite.  Apalache produces a static type
error, if the type of elements of `S` is not compatible in the context of `P`
when an element of `S` is bound to `x`.

**Advanced syntax:** Instead of a single variable `x`, you can use the tuple
syntax to unpack variables from a Cartesian product, see [Tuples](./tuples.md).
For instance, one can write `\E <<x, y>> \in S: P`. In this case, for every
element `e` of `S`, the variable `x` is bound to `e[1]` and `y` is bound to
`e[2]`. The predicate `P` is evaluated against this binding.

Moreover, instead of introducing one variable, one can quantify over several
sets. For instance, you can write: `\E x \in S, y \in T: P`. This form is
simply syntax sugar for the form with nested quantifiers: `\E x \in S: \E y

> \in T: P`.

Reviewer: See comment on `\E`

**Notation:** `e_1 = e_2`

**LaTeX notation:** ![eq](./img/eq.png)

**Arguments:** Two arguments.

**Effect:** This operator evaluates to a Boolean value. It tests the values
of `e_1` and `e_2` for structural equality. The exact effect depends on the
values of `e_1` and `e_2`. Let `e_1` and `e_2` evaluate to the values
`v_1` and `v_2`. Then `e_1 = e_2` evaluates to:

 - If `v_1` and `v_2` are Booleans, then `e_1 = e_2` evaluates to `v_1 <=>
   v_2`.

 - If `v_1` and `v_2` are integers, then `e_1 = e_2` evaluates to `TRUE`
   if and only if `v_1` and `v_2` are exactly the same integers.

 - If `v_1` and `v_2` are strings, then `e_1 = e_2` evaluates to `TRUE`
   if and only if `v_1` and `v_2` are exactly the same strings.

 - If `v_1` and `v_2` are sets, then `e_1 = e_2` evaluates to `TRUE`
   if and only if the following expression evaluates to `TRUE`:

    ```tla
    v_1 \subseteq v_2 /\ v_2 \subseteq v_1
    ```

 - If `v_1` and `v_2` are functions, tuples, records, or sequences,
    then `e_1 = e_2` evaluates to `TRUE`
    if and only if the following expression evaluates to `TRUE`:

    ```tla
      DOMAIN v_1 = DOMAIN v_2 /\ \A x \in DOMAIN v_1: v_1[x] = v_2[x]
    ```

>  - In other cases, `e_1 = e_2` evaluates to `FALSE`. TLC and Apalache report an
>    error, as the values have incomparable types.

Reviewer 2: So this be

```suggestion
 - In other cases, `e_1 = e_2` evaluates to `FALSE` if the values have comparable types.
 - TLC and Apalache report an error, if the values have incomparable types.
```

   
**Determinism:** Deterministic, unless `e_1` has the form `x'`, which can be

> interpreted an assignment to the variable `x'`.  For the non-deterministic

Reviewer 2: ```suggestion
interpreted as an assignment to the variable `x'`.  For the non-deterministic
```

case, see [Control Flow and Non-determinism].

**Errors:** Pure TLA+ does not restrict the operator arguments. TLC flags a

> model checking error, `e_1` and `e_2` evaluate to incomparable values.

Reviewer: ```suggestion
model checking error, if `e_1` and `e_2` evaluate to incomparable values.
```

> Apalache produces a static type error, if the type of `e_1` and `e_2` do not

Reviewer: ```suggestion
Apalache produces a static type error, if the types of `e_1` and `e_2` do not
```

  frozenset({ 1, 2 }) == frozenset({ 2, 1 })
  frozenset({ 1 }) - frozenset({ 1 }) == frozenset({ "a" }) - frozenset({ "a" })
  frozenset({ frozenset({ 1, 2 }) }) == frozenset({ frozenset({ 1, 2, 2, 2 }) })
  (1, "a") == (1, "a")
  (1, "a") == (1, "b")
  (1, False) == (2, )
  (1, 2) == (1, 2, 3)
  { "a": 1, "b": 3 } == { "a": 1, "b": 3 }
  { "a": 1, "b": 3 } == { "a": 1 }
  { x: (x + x) for x in { 2 } } == { x: (x * x) for x in { 2 } }
  { x: (x + x) for x in { 2, 3 } } == { x: 2 * x for x in { 2, 3 } } 
```

----------------------------------------------------------------------------

<a name="neq"></a>
### Inequality

**Notation:** `e_1 /= e_2` or `e_1 # e_2`

**LaTeX notation:** ![ne](./img/ne.png)

**Arguments:** Two arguments.

**Effect:** This operator is syntax sugar for `~(e_1 = e_2)`. Full stop.

----------------------------------------------------------------------------

<a name="chooseBounded"></a>
### Bounded Choice

_This operator causes a lot of confusion. Read carefully!_

**Notation:** `CHOOSE x \in S: P`

**LaTeX notation:** ![choose-bounded](./img/choose-bounded.png)

**Arguments:** Three arguments: a variable name, a set, and an
expression.

> **Effect:** This operator is a magical device by the definition, no kidding. It

Reviewer 2: I don't understand this sentence.

Reviewer: Agreed, tone down the theatrics in a manual.

is a blackbox algorithm that _somehow_ picks one element `e` of `S` that
satisfies the expression `P` against the binding `[x |-> e]` and returns this
element `e`.

Did we say "an algorithm"? Yes! `CHOOSE x \in S: P` is deterministic.  If you
give it two equal sets and two equivalent predicates, it will give you the
same value. More formally, the only known property of `CHOOSE` is as follows:

```tla
  {x \in S: P} = {y \in T: Q} =>
      (CHOOSE x \in S: P) = (CHOOSE y \in T: Q)

> ```

Reviewer: Is this really true? Take `S == {1}` and `T == {1,2}`, then `P == TRUE` and `Q == y = 1`. Obviously `S /= T` and `~(P <=> Q) `, even though `{x\ in {1}: TRUE} = {y \in {1,2}: y = 1}`. 
My understanding is that for `P <=> Q` it holds that `(CHOOSE x \in S: P) = (CHOOSE x \in S: Q)` and for `S = T`, then `(CHOOSE x \in S: P) = (CHOOSE x \in T: P)`, but I'm not sure you can combine those two properties.

Author: The book defines `CHOOSE x \in S` as `CHOOSE (x \in S) /\ P`, so it should be true. I am not sure about TLC though. Of course, what is written in English is slightly more restrictive.

Importantly, when `{x \in S: P} = {}`, the expression `CHOOSE x \ in S: P`
    evaluates to an undefined value.

How does it work? We don't know. Nobody knows. Maybe it returns the first

> element of the set? Well, sets are not ordered, so there is no first element.

Reviewer: Again, a bit theatrical.

Author: Fixed

> Why shall you use it? Actually, you should not. Unless you have no other

Reviewer: ```suggestion
Why should you use it? Actually, you should not. Unless you have no other
```

    the only element of `{x \in S: P}`. No magic.
    For instance, see: [Min and Max in FiniteSetsExt].

 - _Use case 2:_ Enumerating set elements in a fixed but unknown order.
    For instance, see: [ReduceSet in FiniteSetsExt].

In other cases, we believe that `CHOOSE` is bound to do [Program synthesis].
So TLC does some form of synthesis by brute force when it has to evaluate
`CHOOSE`.
   
**Determinism:** Deterministic. Very much deterministic. Don't try to model
non-determinism with `CHOOSE`. For non-determinism, see:
[Control Flow and Non-determinism].

Apalache picks a set element that satisfies the predicate `P`, but it does not
guarantee the repeatability property of CHOOSE. It does not guarantee
non-determinism either. Interestingly, this behavior does not really make a
difference for the use cases 1 and 2. If you believe that this causes a problem
in your specification, [open an issue...]

**Errors:** Pure TLA+ does not restrict the operator arguments.  TLC flags a
model checking error, if `S` is infinite.  Apalache produces a static type
error, if the type of elements of `S` is not compatible in the context of `P`

> when an element of `S` is bound to `x`.

Reviewer: See previous.

**Example in TLA+:**

```tla
  CHOOSE x \in 1..3: x >= 3
  \* 3
  CHOOSE x \in 1..3:
    \A y \in 1..3: y >= x
  \* 1, the minimum
  CHOOSE f \in [ 1..10 -> BOOLEAN ]:
    \E x, y \in DOMAIN f:
      f[x] /\ ~f[y]
  \* some Boolean function from 1..10 that covers FALSE and TRUE
```

**Example in Python:** Python does not have anything similar to `CHOOSE`.
The closest possible solution is to sort the filtered set by the string values
and pick the first one (or the last one). So we have introduced a particular

> way of implementing CHOOSE, see [choose.py](./examples/choose.py):

Reviewer: You can do:
```python
def get(s):
  for e in s:
    return e

print(get({"1", 1, (1,)}))
```
under the hood, set is iterable, so you get the 1st element of that iteration.

Author: Yes, but that would not guarantee the determinism of CHOOSE. I am enforcing determinism by sorting the elements with their string representation as a key. It is quite inefficient, but sound.

**Effect:** This operator evaluates to a Boolean value. It evaluates to `TRUE`,
when at least one element in the logical universe makes the expression `P`
evaluate to `TRUE` against the binding `[x |-> e]`. More precisely, we have to
consider only the elements that produced a defined result when evaluating `P`.

Neither TLC, nor Apalache support this operator. It is impossible to give
operational semantics for this operator, unless we explicitly introduce the
universe. It requires a first-order logic solver. This operator may be useful
when writing proofs with [TLAPS].

----------------------------------------------------------------------------

<a name="choose"></a>
### Unbounded CHOOSE

**Notation:** `CHOOSE x: P`

**LaTeX notation:** CHOOSE x: P

**Arguments:** At least two arguments: a variable name and an
expression. 

**Effect:** This operator evaluates to some value `v` in the logical universe
that evaluates `P` to `TRUE` against the binding `[x |-> v]`.

Neither TLC, nor Apalache support this operator. It is impossible to give
operational semantics for this operator, unless we explicitly introduce the
universe and introduce a fixed rule for enumerating its elements.

> Congratulations! You have reached the bottom of this page. If you like to learn

Reviewer: ```suggestion
Congratulations! You have reached the bottom of this page. If you want to learn
```
